frame/treasury/src/lib.rs

```diff
@@ -175,92 +128,182 @@ pub struct Proposal<AccountId, Balance> {
+		type RejectOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The overarching event type.
+		type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Handler for the unbalanced decrease when slashing for a rejected proposal or bounty.
+		type OnSlash: OnUnbalanced<NegativeImbalanceOf<Self, I>>;
+
+		/// Fraction of a proposal's value that should be bonded in order to place the proposal.
+		/// An accepted proposal gets these back. A rejected proposal does not.
+		#[pallet::constant]
+		type ProposalBond: Get<Permill>;
+
+		/// Minimum amount of funds that should be placed in a deposit for making a proposal.
+		#[pallet::constant]
+		type ProposalBondMinimum: Get<BalanceOf<Self, I>>;
+
+		/// Period between successive spends.
+		#[pallet::constant]
+		type SpendPeriod: Get<Self::BlockNumber>;
+
+		/// Percentage of spare funds (if any) that are burnt per spend period.
+		#[pallet::constant]
+		type Burn: Get<Permill>;
+
+		/// The treasury's module id, used for deriving its sovereign account ID.
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Handler for the unbalanced decrease when treasury funds are burned.
+		type BurnDestination: OnUnbalanced<NegativeImbalanceOf<Self, I>>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
 
-		/// Proposals that have been made.
-		pub Proposals get(fn proposals):
-			map hasher(twox_64_concat) ProposalIndex
-			=> Option<Proposal<T::AccountId, BalanceOf<T, I>>>;
+		/// Runtime hooks to external pallet using treasury to compute spend funds.
+		type SpendFunds: SpendFunds<Self, I>;
 
-		/// Proposal indices that have been approved but not yet awarded.
-		pub Approvals get(fn approvals): BoundedVec<ProposalIndex, T::MaxApprovals>;
+		/// The maximum number of approvals that can wait in the spending queue.
+		type MaxApprovals: Get<u32>;
+	}
+
+	/// Number of proposals that have been made.
+	#[pallet::storage]
+	#[pallet::getter(fn proposal_count)]
+	pub(crate) type ProposalCount<T, I = ()> = StorageValue<_, ProposalIndex, ValueQuery>;
+
+	/// Proposals that have been made.
+	#[pallet::storage]
+	#[pallet::getter(fn proposals)]
+	pub type Proposals<T: Config<I>, I: 'static = ()> = StorageMap<
+		_,
+		Twox64Concat,
+		ProposalIndex,
+		Proposal<T::AccountId, BalanceOf<T, I>>,
+		OptionQuery
+	>;
+
+	/// Proposal indices that have been approved but not yet awarded.
+	#[pallet::storage]
+	#[pallet::getter(fn approvals)]
+	pub type Approvals<T: Config<I>, I: 'static = ()> = StorageValue<
+		_,
+		BoundedVec<ProposalIndex, T::MaxApprovals>,
+		ValueQuery
+	>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig;
+
+	#[cfg(feature = "std")]
+	impl Default for GenesisConfig {
+		fn default() -> Self {
+			Self
+		}
+	}
+
+	#[cfg(feature = "std")]
+	impl GenesisConfig {
+		/// Direct implementation of `GenesisBuild::assimilate_storage`.
+		pub fn assimilate_storage<T: Config<I>, I: 'static>(
```

Comment:
I think we should deprecate this direct implementation, I know it was generated by decl_storage, but people shouldn't really need it if they use construct_runtime and the runtime GensisConfig build.
```suggestion
		#[deprecated(note = "use `<GensisConfig<T, I> as GenesisBuild<T, I>>::assimilate_storage` instead")]
		pub fn assimilate_storage<T: Config<I>, I: 'static>(
```